src/Atc.Kepware.Configuration/Services/Connectivity/KepwareConfigurationClientConnectivity.cs

```diff
@@ -472,6 +502,95 @@ public sealed partial class KepwareConfigurationClient
+                Exception = response.Exception,
+            };
+        }
+
+        // No actual devices, return early, nothing to adapt
+        if (response.Data!.Count == 0)
+        {
+            return new HttpClientRequestResult<IList<TDevice>?>([])
+            {
+                CommunicationSucceeded = response.CommunicationSucceeded,
+                StatusCode = response.StatusCode,
+                Message = response.Message,
+                Exception = response.Exception,
+            };
+        }
+
+        if (!TryGetDeviceTypeJsonMappingType<TDevice>(out Type? jsonMappingType))
+        {
+            return new HttpClientRequestResult<IList<TDevice>?>()
+            {
+                CommunicationSucceeded = response.CommunicationSucceeded,
+                StatusCode = response.StatusCode,
+                Message = response.Message,
+                Exception = new NotSupportedException($"Could not find a JSON mapping type for {typeof(TDevice).Name}"),
+            };
+        }
+
+        // Serialize to JSON mapping type
+        IList<object?> mappedDeviceType = response
+            .Data
+            .Select(x => JsonSerializer.Deserialize(x.ToString(), jsonMappingType))
+            .ToList();
+
+        // Adapt to the desired type
+        Type deviceTypeListType = typeof(IList<>).MakeGenericType(typeof(TDevice));
+        IList<TDevice>? deviceTypes = (IList<TDevice>?)mappedDeviceType.Adapt(typeof(IList<object?>), deviceTypeListType);
+
+        return new HttpClientRequestResult<IList<TDevice>?>(deviceTypes)
+        {
+            CommunicationSucceeded = response.CommunicationSucceeded,
+            StatusCode = response.StatusCode,
+            Message = response.Message,
+        };
+    }
+
+    private static bool TryGetDeviceTypeJsonMappingType<TDevice>([NotNullWhen(true)] out Type? jsonMappingType)
+        where TDevice : DeviceBase
+    {
+        jsonMappingType = DeviceTypeJsonMappingTypeLookup.GetOrAdd(typeof(TDevice), GetDeviceTypeJsonMappingType);
+        return jsonMappingType != null;
+    }
+
+    /// <summary>
+    /// Each driver device type has a corresponding type in this assembly for mapping from JSON. Try finding
+    /// that type by looking for a shared interface with the device type, that derives from IDeviceBase.
+    /// </summary>
+    /// <returns>
+    /// The type that maps to the device type, or null if not found.
+    /// </returns>
+    private static Type? GetDeviceTypeJsonMappingType(Type deviceType)
+    {
+        // Get the implemented interfaces that derive from IDeviceBase, but are not IDeviceBase
+        ISet<Type> implementedInterfaces = deviceType
+            .GetInterfaces()
+            .Where(x
+                => typeof(IDeviceBase).IsAssignableFrom(x)
+                && x != typeof(IDeviceBase))
+            .ToHashSet();
+
+        // Find the first type in this assembly that shares an interface
+        return Assembly
+            .GetExecutingAssembly()
```

Comment:
*nit: This could cause issues down the line - would be better to utilize GetAssembly(typeof(IAssemblyMarker)) from the specific Project were looking in. Currently no IAssemblyMarker exists.

Reply:
What would be against using `GetAssembly(typeof(KepwareConfigurationClient))`?

Reply:
That would work fine aswell :)